<html> 
	<head> 
		<META NAME="date" CONTENT="2009-12-31"> 
		<META NAME="author" CONTENT="Osnat Penn,Eyal Privman,Haim Ashkenazy,Giddy Landan,Dan Graur,Tal Pupko"> 
		<META NAME="keywords" CONTENT="Multiple sequence alignment, guide tree, alignment confidence,Guidance server"> 
		<META NAME="description" CONTENT="GUIDANCE: a web server for assessing alignment confidence score"> 
		<meta http-equiv="X-UA-Compatible" content="IE=EmulateIE7"> 
<title>GUIDANCE SOURCE CODE</title>
		<link rel="stylesheet" type="text/css" href="guidance.css" />
	</head>
	<body onLoad="init()" bgcolor="#F2F2F2"  align=left  marginheight=0 marginwidth=0 topmargin=0 leftmargin=0 rightmargin=0> 
		<table cols=2 cellspacing=0 cellpadding=10  width=750 border=0> 
		   <tr align=left> 
			   <td width=125 ALIGN= center VALIGN=top bgcolor="#99CCFF"> 
				   <script language="javascript" src="title.js"> </script> 
		   </td> 
 
   <td width=825 ALIGN=left VALIGN=top>

<h1 align=center><font face="Comic Sans MS" size="6" COLOR="#003399">
Source Code</font> <br>
</h1>
<font face=Verdana size=2>
<br><b>
Please note that the GUIDANCE source code is freely distributed for academic use only (see <a href="source.html#copyrights">copyrights</a> statement below)</b><br><br>
</font>
<br>
1. <a href="source.html#download">Downloading and compiling GUIDANCE.</a><br>
2. <a href="source.html#run">Running GUIDANCE<br></a><br>


<a NAME=download><font face=Verdana size=4><b><u>Downloading and Compiling GUIDANCE</u></b></font></a><br>

The following instructions should work right out of the box for
UNIX-like systems.  Mac should also work in principle, but is not yet
supported.  Windows will require some additional work, such as setting
up a cygwin environment.<br>
<br>

1. Download the <a href="guidance.v1.01.tgz">GUIDANCE source code</a>. <br>

2. Unzip and untar the files:<br>
<p>
<blockquote>
tar -xzvf guidance.v1.01.tgz
</blockquote>
</p>
This will create a directory named guidance.v1.01<br>
<br>
3. Make: <br>
<p>
<blockquote>
cd guidance.v1.01<br>
make
</blockquote>
</p>
4. Check if you have the desired alignment program(s) installed:<br>
<ul>
<li><b>MAFFT:</b> Type "mafft" and check that you have version 6.712 or newer.</li>
<ul><li>Else download and install MAFFT from:<br>
<a href="http://mafft.cbrc.jp/alignment/software/">http://mafft.cbrc.jp/alignment/software/</a></li></ul>
<li><b>PRANK:</b> Type "prank" and check that you have version v.100223 or newer.</li>
<ul><LI>Else download and install PRANK from:<br>
<a href="http://www.ebi.ac.uk/goldman-srv/prank/prank/">http://www.ebi.ac.uk/goldman-srv/prank/prank/</a></li></ul>
<li><B>CLUSTALW:</B> Type "clustalw" and check that you have it insalled<br>
<ul><li>Else download and install CLUSTALW from:<br>
<a href="http://www.ebi.ac.uk/Tools/clustalw2/index.html">http://www.ebi.ac.uk/Tools/clustalw2/index.html</a></li></ul><br>
</ul>
5. GUIDANCE also uses Perl and BioPerl:
<ul>
<li>Type "perl -v" and check that you Perl installed.</li>
<ul><li>Else download and install it from:<br>
<a href="http://www.perl.org/">http://www.perl.org/</a></li></ul>
<li>Type "perl -e 'use Bio::SeqIO'" to check that you have BioPerl.</li>
<ul><li>Else download and install it from:<br>
<a href="http://www.bioperl.org/">http://www.bioperl.org/</a></li></ul>
</ul>

<br>
<b>For any problems - please <a
href=mailto:bioSequence@tauex.tau.ac.il>contact us</a><br><br></b>


<a NAME=run><font face=Verdana size=4><b><u>Running GUIDANCE</u></b></font></a><br>
Run the Perl script: guidance/www/Guidance/run_calc.pl<br>
(Note that you cannot move this script out of its directory, because it
uses relative paths to other files in other directories.  Sorry)<br>
GUIDANCE uses flags in the command line arguments: (for help, type:
"guidance")<br>
<br>
<b>USAGE:</b> guidance --seqFile SEQFILE --msaProgram
[MAFFT|PRANK|CLUSTALW] --seqType [aa|nuc|codon] --outDir OUTDIR<br>
<br>
<b>Required parameters:</b><br>
  <b>--seqFile:</b> Input sequence file in FASTA format<br>
  <b>--msaProgram:</b> Which MSA program to use<br>
  <b>--seqType:</b> Type of sequences for alignment (amino acids,
  nucleotides, or codons)<br>
  <b>--outDir:</b> Output directory that will be created automatically
  and hold all output files<br>
<br>
<b>Optional parameters:</b><br>
  <b>--program</b> [GUIDANCE|HoT] Default=GUIDANCE<br>
  <b>--bootstraps:</b> Number of bootstrap iterations (only for
  GUIDQANCE). Defaut=100<br>
  <b>--genCode:</b> Genetic code identifier (only for codon
  sequences). Default=1<br>
                1) Nuclear Standard<br>
                15) Nuclear Blepharisma<br>
                6)  Nuclear Ciliate<br>
                10) Nuclear Euplotid<br>
                2)  Mitochondria Vertebrate<br>
                5)  Mitochondria Invertebrate<br>
                3)  Mitochondria Yeast<br>
                13) Mitochondria Ascidian<br>
                9)  Mitochondria Echinoderm<br>
                14) Mitochondria Flatworm<br>
                4)  Mitochondria Protozoan<br>
  <b>--outOrder</b> [aligned|as_input] default=aligned<br>
  <b>--msaFile:</b> Input alignment file - not recommended, see
  the <a href="http://guidance.tau.ac.il/overview.html#InputMSA">overview</a> section<br>
  <b>--seqCutoff:</b> Confidence cutoff between 0 to 1. Default=0.6<br>
  <b>--colCutoff:</b> Confidence cutoff between 0 to 1. Default=0.93<br>
  <b>--mafft:</b> path to mafft executable. Default=mafft<br>
  <b>--prank:</b> path to prank executable. Default=prank<br>
<br>
<font face="Verdana" size="3" color="red"><u><b>EXAMPLES:</b></u></font> 
				 <ul> 
				 <li><b>guidance --seqFile protein.fas --msaProgram MAFFT --seqType aa --outDir protein.guidance</b><br> 
				 - will align the amino acid sequences in the fasta file &quot;protein.fas&quot; using MAFFT and output all results to the diretory &quot;protein.guidance&quot; </li> 
				 <br> 
				 <li><b>guidance --seqFile codingSeq.fas --msaProgram PRANK --seqType codon --outDir codingSeq.guidance --genCode 2 --bootstraps 30</b><br> 
				 - will align the codon sequences in the fasta file &quot;codingSeq.fas&quot; using PRANK after translation using the vertebrate mitochondrial genetic code and output all results to the diretory &quot;codingSeq.guidance&quot;. Only 30 bootstrap iterations will be done instead of the default 100 (cut run-time by a factor of 3) </li> 
				 <br> 
				 </ul> 
<br>
<center>
<font face="Comic Sans MS" COLOR="#91219E">For any problems or questions
please contact us at <a href=mailto:bioSequence@tauex.tau.ac.il>
bioSequence@tauex.tau.ac.il.</a><br>Enjoy!</font><br><br></center>
</font>


<hr>
<a NAME=copyrights><font face=Verdana size="4" color='red'><b><u>Copyrights:</u></b></font></a><br>
<ul>
<font face=Verdana size=2>
<li><b>To modify the code, or use parts of it for other purposes,
permission should be requested. Please contact <a
href=mailto:talp@post.tau.ac.il>Tal Pupko</a><br></b>
<br></li>
<li><b>Please note that the use of the GUIDANCE program is for academic
use only<br></b>
</font></li>
</ul></FONT>
<hr><br><br>


</table>
</body>

</html>
